lib/sqeduler/lock_maintainer.rb

```diff
@@ -0,0 +1,84 @@
+module Sqeduler
+  # This is to ensure that if you set your jobs to run one a time and something goes wrong
+  # causing a job to run for a long time, your lock won't expire.
+  # This does't stop long running jobs, it just ensures you only end up with one long running job
+  # rather than 20 of them.
+  class LockMaintainer
+    RUN_INTERVAL = 30
+
+    def initialize
+      @class_with_locks = {}
+    end
+
+    # This is only done when we initialize Sqeduler, don't need to worry about threading
+    def run
+      @maintainer_thread ||= Thread.new do
+        loop do
+          begin
+            synchronize
+          rescue => ex
+            Service.logger.error "[MAINTAINER] Error while trying to refres locks, #{ex.class}, #{ex.message}"
+          end
+
+          sleep RUN_INTERVAL
+        end
+      end
+    end
+
+    private
+
+    def synchronize
+      # Not great, but finding our identity in Sidekiq is a pain, and we already have locks in Sqeduler.
+      # Easier to just try and grab a lock each time and whichever server wins gets to do it.
+      return unless redis_lock.take_lock
+
+      now = Time.now.utc.to_i
+
+      Service.redis_pool do |redis|
+        redis.pipelined do
+          workers.each do |_worker, _tid, args|
+            # No sense in pinging if it's not been running long enough to matter
+            next if (now - args["run_at"]) < RUN_INTERVAL
+
+            klass = exclusive_lock_class(args["payload"]["class"])
+            next unless klass
+
+            lock_key = klass.sync_lock_key(*args["payload"]["args"])
+
+            # This works because EXPIRE does not recreate the key, it only resets the expiration.
+            # We don't have to worry about atomic operations or anything like that.
+            # If the job finishes in the interim and deletes the key nothing will happen.
+            redis.expire(lock_key, klass.synchronize_jobs_expiration)
```

Comment:
The window for that is small enough that it means we're just adding a couple of seconds to the lock time. It won't cause any issues because as soon as the job finishes, it deletes the key anyway.

The only time it would cause issues is if the job died and couldn't release the lock. Since at best, this only adds a few extra seconds to the locking time, it's a tradeoff I think that's worth it.